*analyzing indels generated by CRISPR or other nucleases via deep sequencing*
====================

ipython notebook detailing the procedure to map sequencing reads to a small 'genome' and analyzing these alignments for insertions and deletions

software requirements and *tested versions*:<br>
  • gmap/gsnap [http://research-pub.gene.com/gmap/] *version 08-20-2014*<br>
  • pysam [https://github.com/pysam-developers/pysam] *version 0.8.0*<br>
  • samtools/htslib [http://www.htslib.org] *version 1.0*<br>
  
python package requirements:<br>
  • pysamstats [https://github.com/alimanfoo/pysamstats] *version 0.16*<br>
  
